src/MultiMessageReceiver.sol

```diff
@@ -138,57 +182,63 @@ contract MultiMessageReceiver is IMultiMessageReceiver, ExecutorAware, Initializ
         external
         onlySelf
     {
-        require(_receiverAdapters.length == _operations.length, "mismatch length");
-        for (uint256 i; i < _receiverAdapters.length; ++i) {
+        uint256 len = _receiverAdapters.length;
+
+        if (len != _operations.length) {
+            revert Error.ARRAY_LENGTH_MISMATCHED();
+        }
+
+        for (uint256 i; i < len;) {
             _updateReceiverAdapter(_receiverAdapters[i], _operations[i]);
+
+            unchecked {
+                ++i;
+            }
         }
     }
 
     /// @notice Update power quorum threshold of message execution.
-    function updateQuorumThreshold(uint64 _quorumThreshold) external onlySelf {
-        require(_quorumThreshold <= trustedExecutor.length && _quorumThreshold > 0, "invalid threshold");
-        quorumThreshold = _quorumThreshold;
-        emit QuorumThresholdUpdated(_quorumThreshold);
+    function updatequorum(uint64 _quorum) external onlySelf {
+        /// NOTE: should check 2/3 ?
+        if (_quorum > trustedExecutor.length || _quorum == 0) {
+            revert Error.INVALID_QUORUM_THRESHOLD();
+        }
+
+        quorum = _quorum;
+        emit quorumUpdated(_quorum);
     }
 
-    /// @notice View message info, return (executed, msgPower, delivered adapters)
-    function getMessageInfo(bytes32 msgId) public view returns (bool, uint64, address[] memory) {
-        MsgInfo storage msgInfo = msgInfos[msgId];
-        uint64 msgPower = _computeMessagePower(msgInfo);
-        address[] memory adapters = new address[](msgPower);
-        if (msgPower > 0) {
-            uint32 n = 0;
-            for (uint64 i = 0; i < trustedExecutor.length; i++) {
-                address adapter = trustedExecutor[i];
-                if (msgInfo.from[adapter]) {
-                    adapters[n++] = adapter;
-                }
+    /// @notice view message info, return (executed, msgPower, delivered adapters)
+    function getMessageInfo(bytes32 msgId) public view returns (bool, uint256, string[] memory) {
+        uint256 msgCurrentQuorum = messageQuorum[msgId];
+        string[] memory successfulBridge = new string[](msgCurrentQuorum);
+
+        for(uint256 i; i < trustedExecutor.length; ) {
+            if(isDuplicateAdapter[msgId][trustedExecutor[i]]) {
+               successfulBridge[i] = IBridgeReceiverAdapter(trustedExecutor[i]).name();
+            }    
+
+            unchecked {
+                ++i;
```

Comment:
I am not sure this will work. It will likely cause an index out-of-bounds issues because a different index than `i` should keep track of what the next `successfulBridge[]` element should be that should be populated (i.e. not `successfulBridge[i] =...`)